vault/mount.go

```diff
@@ -1286,15 +1469,39 @@ func (c *Core) runMountUpdates(ctx context.Context, needPersist bool) error {
 	}
 
 	// Persist both mount tables
-	if err := c.persistMounts(ctx, c.mounts, nil); err != nil {
+	if err := c.persistMounts(ctx, barrier, c.mounts, nil, ""); err != nil {
 		c.logger.Error("failed to persist mount table", "error", err)
 		return errLoadMountsFailed
 	}
 	return nil
 }
 
-// persistMounts is used to persist the mount table after modification
-func (c *Core) persistMounts(ctx context.Context, table *MountTable, local *bool) error {
+// persistMounts is used to persist the mount table after modification.
+func (c *Core) persistMounts(ctx context.Context, barrier logical.Storage, table *MountTable, local *bool, mount string) error {
+	// Sometimes we may not want to explicitly pass barrier; fetch it if
+	// necessary.
+	if barrier == nil {
+		barrier = c.barrier
+	}
+
+	// Gracefully handle a transaction-aware backend, if a transaction
+	// wasn't created for us. This is safe as we do not support nested
+	// transactions.
+	needTxnCommit := false
+	if _, ok := barrier.(logical.TransactionalStorage); ok {
+		var err error
+		barrier, err = barrier.(logical.TransactionalStorage).BeginTx(ctx)
```

Comment:
nit: assertion safety

Reply:
As above, this is guarded by two lines above if I'm not mistaken. :-)